{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.10.12","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kaggle":{"accelerator":"none","dataSources":[{"sourceId":10891168,"sourceType":"datasetVersion","datasetId":6768117}],"dockerImageVersionId":30918,"isInternetEnabled":true,"language":"python","sourceType":"notebook","isGpuEnabled":false}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:48.930517Z","iopub.execute_input":"2025-03-03T00:35:48.930827Z","iopub.status.idle":"2025-03-03T00:35:49.304429Z","shell.execute_reply.started":"2025-03-03T00:35:48.930799Z","shell.execute_reply":"2025-03-03T00:35:49.303353Z"}},"outputs":[{"name":"stdout","text":"/kaggle/input/gdsc-data-extra/gdsc_data_extra.csv\n","output_type":"stream"}],"execution_count":1},{"cell_type":"code","source":"import pandas as pd\nimport numpy as np\nimport matplotlib.pyplot as plt\nfrom sklearn.preprocessing import StandardScaler\nfrom sklearn.cluster import KMeans, DBSCAN, MiniBatchKMeans\nfrom sklearn.metrics import silhouette_score\nfrom scipy.spatial.distance import pdist, squareform\nfrom scipy.cluster.hierarchy import linkage, dendrogram, fcluster\nimport time\nfrom matplotlib.cm import viridis\nimport warnings\nwarnings.filterwarnings('ignore')","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:49.306558Z","iopub.execute_input":"2025-03-03T00:35:49.306924Z","iopub.status.idle":"2025-03-03T00:35:49.889097Z","shell.execute_reply.started":"2025-03-03T00:35:49.306900Z","shell.execute_reply":"2025-03-03T00:35:49.888092Z"}},"outputs":[],"execution_count":2},{"cell_type":"code","source":"pip install kneed","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:49.890625Z","iopub.execute_input":"2025-03-03T00:35:49.891138Z","iopub.status.idle":"2025-03-03T00:35:53.906372Z","shell.execute_reply.started":"2025-03-03T00:35:49.891101Z","shell.execute_reply":"2025-03-03T00:35:53.905347Z"}},"outputs":[{"name":"stdout","text":"Requirement already satisfied: kneed in /usr/local/lib/python3.10/dist-packages (0.8.5)\nRequirement already satisfied: numpy>=1.14.2 in /usr/local/lib/python3.10/dist-packages (from kneed) (1.26.4)\nRequirement already satisfied: scipy>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from kneed) (1.13.1)\nRequirement already satisfied: mkl_fft in /usr/local/lib/python3.10/dist-packages (from numpy>=1.14.2->kneed) (1.3.8)\nRequirement already satisfied: mkl_random in /usr/local/lib/python3.10/dist-packages (from numpy>=1.14.2->kneed) (1.2.4)\nRequirement already satisfied: mkl_umath in /usr/local/lib/python3.10/dist-packages (from numpy>=1.14.2->kneed) (0.1.1)\nRequirement already satisfied: mkl in /usr/local/lib/python3.10/dist-packages (from numpy>=1.14.2->kneed) (2025.0.1)\nRequirement already satisfied: tbb4py in /usr/local/lib/python3.10/dist-packages (from numpy>=1.14.2->kneed) (2022.0.0)\nRequirement already satisfied: mkl-service in /usr/local/lib/python3.10/dist-packages (from numpy>=1.14.2->kneed) (2.4.1)\nRequirement already satisfied: intel-openmp>=2024 in /usr/local/lib/python3.10/dist-packages (from mkl->numpy>=1.14.2->kneed) (2024.2.0)\nRequirement already satisfied: tbb==2022.* in /usr/local/lib/python3.10/dist-packages (from mkl->numpy>=1.14.2->kneed) (2022.0.0)\nRequirement already satisfied: tcmlib==1.* in /usr/local/lib/python3.10/dist-packages (from tbb==2022.*->mkl->numpy>=1.14.2->kneed) (1.2.0)\nRequirement already satisfied: intel-cmplr-lib-rt in /usr/local/lib/python3.10/dist-packages (from mkl_umath->numpy>=1.14.2->kneed) (2024.2.0)\nRequirement already satisfied: intel-cmplr-lib-ur==2024.2.0 in /usr/local/lib/python3.10/dist-packages (from intel-openmp>=2024->mkl->numpy>=1.14.2->kneed) (2024.2.0)\nNote: you may need to restart the kernel to use updated packages.\n","output_type":"stream"}],"execution_count":3},{"cell_type":"code","source":"gdsc_data = pd.read_csv('/kaggle/input/gdsc-data-extra/gdsc_data_extra.csv') \nprint(f\"Dataset shape: {gdsc_data.shape}\")\ngdsc_data.head().T","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:53.907534Z","iopub.execute_input":"2025-03-03T00:35:53.907889Z","iopub.status.idle":"2025-03-03T00:35:54.636786Z","shell.execute_reply.started":"2025-03-03T00:35:53.907860Z","shell.execute_reply":"2025-03-03T00:35:54.635757Z"}},"outputs":[{"name":"stdout","text":"Dataset shape: (198342, 21)\n","output_type":"stream"},{"execution_count":4,"output_type":"execute_result","data":{"text/plain":"                                                       0                1  \\\nCOSMIC_ID                                         683667           687448   \nCELL_LINE_NAME                                    PFSK-1         COLO-829   \nTCGA_DESC                                             MB             SKCM   \nDRUG_ID                                             1003             1003   \nDRUG_NAME                                   Camptothecin     Camptothecin   \nLN_IC50                                        -1.463887        -1.235034   \nAUC                                              0.93022         0.867348   \nZ_SCORE                                         0.433123         0.557727   \nGDSC Tissue descriptor 1                  nervous_system             skin   \nGDSC Tissue descriptor 2                 medulloblastoma         melanoma   \nCancer Type (matching TCGA label)                     MB             SKCM   \nMicrosatellite instability Status (MSI)        MSS/MSI-L        MSS/MSI-L   \nScreen Medium                                          R                R   \nGrowth Properties                               Adherent         Adherent   \nCNA                                                    1                1   \nGene Expression                                        1                1   \nMethylation                                            1                1   \nTARGET                                              TOP1             TOP1   \nTARGET_PATHWAY                           DNA replication  DNA replication   \nDrug_Potency_Index                              4.037985         3.558215   \nResponse_Variability                            0.390564         0.448556   \n\n                                                       2                  3  \\\nCOSMIC_ID                                         687452             687455   \nCELL_LINE_NAME                                      5637                RT4   \nTCGA_DESC                                           BLCA               BLCA   \nDRUG_ID                                             1003               1003   \nDRUG_NAME                                   Camptothecin       Camptothecin   \nLN_IC50                                        -2.632632          -2.963191   \nAUC                                             0.834067           0.821438   \nZ_SCORE                                        -0.203221            -0.3832   \nGDSC Tissue descriptor 1                         Unknown  urogenital_system   \nGDSC Tissue descriptor 2                         Unknown            Bladder   \nCancer Type (matching TCGA label)                Unknown               BLCA   \nMicrosatellite instability Status (MSI)        MSS/MSI-L          MSS/MSI-L   \nScreen Medium                                          R              D/F12   \nGrowth Properties                               Adherent           Adherent   \nCNA                                                    1                  1   \nGene Expression                                        1                  1   \nMethylation                                            1                  1   \nTARGET                                              TOP1               TOP1   \nTARGET_PATHWAY                           DNA replication    DNA replication   \nDrug_Potency_Index                              5.236444           5.619105   \nResponse_Variability                            0.262146           0.527685   \n\n                                                         4  \nCOSMIC_ID                                           687457  \nCELL_LINE_NAME                                       SW780  \nTCGA_DESC                                             BLCA  \nDRUG_ID                                               1003  \nDRUG_NAME                                     Camptothecin  \nLN_IC50                                          -1.449138  \nAUC                                                0.90505  \nZ_SCORE                                           0.441154  \nGDSC Tissue descriptor 1                 urogenital_system  \nGDSC Tissue descriptor 2                           Bladder  \nCancer Type (matching TCGA label)                     BLCA  \nMicrosatellite instability Status (MSI)          MSS/MSI-L  \nScreen Medium                                        D/F12  \nGrowth Properties                                 Adherent  \nCNA                                                      1  \nGene Expression                                          1  \nMethylation                                              1  \nTARGET                                                TOP1  \nTARGET_PATHWAY                             DNA replication  \nDrug_Potency_Index                                3.943711  \nResponse_Variability                              0.395158  ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>0</th>\n      <th>1</th>\n      <th>2</th>\n      <th>3</th>\n      <th>4</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>COSMIC_ID</th>\n      <td>683667</td>\n      <td>687448</td>\n      <td>687452</td>\n      <td>687455</td>\n      <td>687457</td>\n    </tr>\n    <tr>\n      <th>CELL_LINE_NAME</th>\n      <td>PFSK-1</td>\n      <td>COLO-829</td>\n      <td>5637</td>\n      <td>RT4</td>\n      <td>SW780</td>\n    </tr>\n    <tr>\n      <th>TCGA_DESC</th>\n      <td>MB</td>\n      <td>SKCM</td>\n      <td>BLCA</td>\n      <td>BLCA</td>\n      <td>BLCA</td>\n    </tr>\n    <tr>\n      <th>DRUG_ID</th>\n      <td>1003</td>\n      <td>1003</td>\n      <td>1003</td>\n      <td>1003</td>\n      <td>1003</td>\n    </tr>\n    <tr>\n      <th>DRUG_NAME</th>\n      <td>Camptothecin</td>\n      <td>Camptothecin</td>\n      <td>Camptothecin</td>\n      <td>Camptothecin</td>\n      <td>Camptothecin</td>\n    </tr>\n    <tr>\n      <th>LN_IC50</th>\n      <td>-1.463887</td>\n      <td>-1.235034</td>\n      <td>-2.632632</td>\n      <td>-2.963191</td>\n      <td>-1.449138</td>\n    </tr>\n    <tr>\n      <th>AUC</th>\n      <td>0.93022</td>\n      <td>0.867348</td>\n      <td>0.834067</td>\n      <td>0.821438</td>\n      <td>0.90505</td>\n    </tr>\n    <tr>\n      <th>Z_SCORE</th>\n      <td>0.433123</td>\n      <td>0.557727</td>\n      <td>-0.203221</td>\n      <td>-0.3832</td>\n      <td>0.441154</td>\n    </tr>\n    <tr>\n      <th>GDSC Tissue descriptor 1</th>\n      <td>nervous_system</td>\n      <td>skin</td>\n      <td>Unknown</td>\n      <td>urogenital_system</td>\n      <td>urogenital_system</td>\n    </tr>\n    <tr>\n      <th>GDSC Tissue descriptor 2</th>\n      <td>medulloblastoma</td>\n      <td>melanoma</td>\n      <td>Unknown</td>\n      <td>Bladder</td>\n      <td>Bladder</td>\n    </tr>\n    <tr>\n      <th>Cancer Type (matching TCGA label)</th>\n      <td>MB</td>\n      <td>SKCM</td>\n      <td>Unknown</td>\n      <td>BLCA</td>\n      <td>BLCA</td>\n    </tr>\n    <tr>\n      <th>Microsatellite instability Status (MSI)</th>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n    </tr>\n    <tr>\n      <th>Screen Medium</th>\n      <td>R</td>\n      <td>R</td>\n      <td>R</td>\n      <td>D/F12</td>\n      <td>D/F12</td>\n    </tr>\n    <tr>\n      <th>Growth Properties</th>\n      <td>Adherent</td>\n      <td>Adherent</td>\n      <td>Adherent</td>\n      <td>Adherent</td>\n      <td>Adherent</td>\n    </tr>\n    <tr>\n      <th>CNA</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>Gene Expression</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>Methylation</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>TARGET</th>\n      <td>TOP1</td>\n      <td>TOP1</td>\n      <td>TOP1</td>\n      <td>TOP1</td>\n      <td>TOP1</td>\n    </tr>\n    <tr>\n      <th>TARGET_PATHWAY</th>\n      <td>DNA replication</td>\n      <td>DNA replication</td>\n      <td>DNA replication</td>\n      <td>DNA replication</td>\n      <td>DNA replication</td>\n    </tr>\n    <tr>\n      <th>Drug_Potency_Index</th>\n      <td>4.037985</td>\n      <td>3.558215</td>\n      <td>5.236444</td>\n      <td>5.619105</td>\n      <td>3.943711</td>\n    </tr>\n    <tr>\n      <th>Response_Variability</th>\n      <td>0.390564</td>\n      <td>0.448556</td>\n      <td>0.262146</td>\n      <td>0.527685</td>\n      <td>0.395158</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}],"execution_count":4},{"cell_type":"code","source":"features = ['AUC', 'Z_SCORE', 'Drug_Potency_Index', 'Response_Variability']\nX = gdsc_data[features]\ny = gdsc_data['LN_IC50']","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:54.637755Z","iopub.execute_input":"2025-03-03T00:35:54.638069Z","iopub.status.idle":"2025-03-03T00:35:54.645195Z","shell.execute_reply.started":"2025-03-03T00:35:54.638043Z","shell.execute_reply":"2025-03-03T00:35:54.644274Z"}},"outputs":[],"execution_count":5},{"cell_type":"code","source":"numerical_features_withLabel = ['LN_IC50', 'AUC', 'Z_SCORE', 'Drug_Potency_Index', 'Response_Variability']\nnumerical_features = gdsc_data[numerical_features_withLabel]\nnumerical_features.to_csv('numerical_features.csv', index=False)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:54.646108Z","iopub.execute_input":"2025-03-03T00:35:54.646398Z","iopub.status.idle":"2025-03-03T00:35:55.767645Z","shell.execute_reply.started":"2025-03-03T00:35:54.646375Z","shell.execute_reply":"2025-03-03T00:35:55.766369Z"}},"outputs":[],"execution_count":6},{"cell_type":"code","source":"numerical_features.head()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:55.768780Z","iopub.execute_input":"2025-03-03T00:35:55.769124Z","iopub.status.idle":"2025-03-03T00:35:55.780338Z","shell.execute_reply.started":"2025-03-03T00:35:55.769095Z","shell.execute_reply":"2025-03-03T00:35:55.779333Z"}},"outputs":[{"execution_count":7,"output_type":"execute_result","data":{"text/plain":"    LN_IC50       AUC   Z_SCORE  Drug_Potency_Index  Response_Variability\n0 -1.463887  0.930220  0.433123            4.037985              0.390564\n1 -1.235034  0.867348  0.557727            3.558215              0.448556\n2 -2.632632  0.834067 -0.203221            5.236444              0.262146\n3 -2.963191  0.821438 -0.383200            5.619105              0.527685\n4 -1.449138  0.905050  0.441154            3.943711              0.395158","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>LN_IC50</th>\n      <th>AUC</th>\n      <th>Z_SCORE</th>\n      <th>Drug_Potency_Index</th>\n      <th>Response_Variability</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>-1.463887</td>\n      <td>0.930220</td>\n      <td>0.433123</td>\n      <td>4.037985</td>\n      <td>0.390564</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>-1.235034</td>\n      <td>0.867348</td>\n      <td>0.557727</td>\n      <td>3.558215</td>\n      <td>0.448556</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>-2.632632</td>\n      <td>0.834067</td>\n      <td>-0.203221</td>\n      <td>5.236444</td>\n      <td>0.262146</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>-2.963191</td>\n      <td>0.821438</td>\n      <td>-0.383200</td>\n      <td>5.619105</td>\n      <td>0.527685</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>-1.449138</td>\n      <td>0.905050</td>\n      <td>0.441154</td>\n      <td>3.943711</td>\n      <td>0.395158</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}],"execution_count":7},{"cell_type":"code","source":"print(\"\\nFeatures for clustering:\")\nprint(X.head())","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:55.783678Z","iopub.execute_input":"2025-03-03T00:35:55.783950Z","iopub.status.idle":"2025-03-03T00:35:55.798420Z","shell.execute_reply.started":"2025-03-03T00:35:55.783925Z","shell.execute_reply":"2025-03-03T00:35:55.797514Z"}},"outputs":[{"name":"stdout","text":"\nFeatures for clustering:\n        AUC   Z_SCORE  Drug_Potency_Index  Response_Variability\n0  0.930220  0.433123            4.037985              0.390564\n1  0.867348  0.557727            3.558215              0.448556\n2  0.834067 -0.203221            5.236444              0.262146\n3  0.821438 -0.383200            5.619105              0.527685\n4  0.905050  0.441154            3.943711              0.395158\n","output_type":"stream"}],"execution_count":8},{"cell_type":"code","source":"scaler = StandardScaler()\nX_scaled = scaler.fit_transform(X)\nX_df = pd.DataFrame(X_scaled, columns=features)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:55.800046Z","iopub.execute_input":"2025-03-03T00:35:55.800305Z","iopub.status.idle":"2025-03-03T00:35:55.829140Z","shell.execute_reply.started":"2025-03-03T00:35:55.800277Z","shell.execute_reply":"2025-03-03T00:35:55.828162Z"}},"outputs":[],"execution_count":9},{"cell_type":"code","source":"n_features = ['LN_IC50', 'AUC', 'Z_SCORE', 'Drug_Potency_Index', 'Response_Variability']\nscaler1 = StandardScaler()\nnumerical_features_scaled = scaler1.fit_transform(numerical_features)\nnumerical_features_scaled_df = pd.DataFrame(numerical_features_scaled, columns=n_features)\nnumerical_features_scaled_df.head()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:55.830311Z","iopub.execute_input":"2025-03-03T00:35:55.830599Z","iopub.status.idle":"2025-03-03T00:35:55.857935Z","shell.execute_reply.started":"2025-03-03T00:35:55.830570Z","shell.execute_reply":"2025-03-03T00:35:55.857099Z"}},"outputs":[{"execution_count":10,"output_type":"execute_result","data":{"text/plain":"    LN_IC50       AUC   Z_SCORE  Drug_Potency_Index  Response_Variability\n0 -2.527836  0.075527  0.360484            2.326585             -0.709719\n1 -2.409852 -0.817682  0.500502            2.092022             -0.647003\n2 -3.130381 -1.290498 -0.354576            2.912521             -0.848600\n3 -3.300801 -1.469916 -0.556818            3.099607             -0.561426\n4 -2.520233 -0.282058  0.369508            2.280494             -0.704752","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>LN_IC50</th>\n      <th>AUC</th>\n      <th>Z_SCORE</th>\n      <th>Drug_Potency_Index</th>\n      <th>Response_Variability</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>-2.527836</td>\n      <td>0.075527</td>\n      <td>0.360484</td>\n      <td>2.326585</td>\n      <td>-0.709719</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>-2.409852</td>\n      <td>-0.817682</td>\n      <td>0.500502</td>\n      <td>2.092022</td>\n      <td>-0.647003</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>-3.130381</td>\n      <td>-1.290498</td>\n      <td>-0.354576</td>\n      <td>2.912521</td>\n      <td>-0.848600</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>-3.300801</td>\n      <td>-1.469916</td>\n      <td>-0.556818</td>\n      <td>3.099607</td>\n      <td>-0.561426</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>-2.520233</td>\n      <td>-0.282058</td>\n      <td>0.369508</td>\n      <td>2.280494</td>\n      <td>-0.704752</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}],"execution_count":10},{"cell_type":"code","source":"numerical_features_scaled_df.to_csv('numerical_features_scaled_df.csv', index=False)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:55.858786Z","iopub.execute_input":"2025-03-03T00:35:55.859042Z","iopub.status.idle":"2025-03-03T00:35:57.403253Z","shell.execute_reply.started":"2025-03-03T00:35:55.859020Z","shell.execute_reply":"2025-03-03T00:35:57.402418Z"}},"outputs":[],"execution_count":11},{"cell_type":"code","source":"X_df.head()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:57.404159Z","iopub.execute_input":"2025-03-03T00:35:57.404427Z","iopub.status.idle":"2025-03-03T00:35:57.414271Z","shell.execute_reply.started":"2025-03-03T00:35:57.404404Z","shell.execute_reply":"2025-03-03T00:35:57.413453Z"}},"outputs":[{"execution_count":12,"output_type":"execute_result","data":{"text/plain":"        AUC   Z_SCORE  Drug_Potency_Index  Response_Variability\n0  0.075527  0.360484            2.326585             -0.709719\n1 -0.817682  0.500502            2.092022             -0.647003\n2 -1.290498 -0.354576            2.912521             -0.848600\n3 -1.469916 -0.556818            3.099607             -0.561426\n4 -0.282058  0.369508            2.280494             -0.704752","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>AUC</th>\n      <th>Z_SCORE</th>\n      <th>Drug_Potency_Index</th>\n      <th>Response_Variability</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>0.075527</td>\n      <td>0.360484</td>\n      <td>2.326585</td>\n      <td>-0.709719</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>-0.817682</td>\n      <td>0.500502</td>\n      <td>2.092022</td>\n      <td>-0.647003</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>-1.290498</td>\n      <td>-0.354576</td>\n      <td>2.912521</td>\n      <td>-0.848600</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>-1.469916</td>\n      <td>-0.556818</td>\n      <td>3.099607</td>\n      <td>-0.561426</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>-0.282058</td>\n      <td>0.369508</td>\n      <td>2.280494</td>\n      <td>-0.704752</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}],"execution_count":12},{"cell_type":"code","source":"norm = plt.Normalize(y.min(), y.max())\ncolors = viridis(norm(y))","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:57.415138Z","iopub.execute_input":"2025-03-03T00:35:57.415376Z","iopub.status.idle":"2025-03-03T00:35:57.435726Z","shell.execute_reply.started":"2025-03-03T00:35:57.415355Z","shell.execute_reply":"2025-03-03T00:35:57.434657Z"}},"outputs":[],"execution_count":13},{"cell_type":"code","source":"def silhouette_analysis_with_minibatch():\n    k_range = range(2, 11)\n    silhouette_avg = []\n    \n    # Sample for silhouette calculation\n    sample_size = min(10000, int(0.05 * len(X_scaled)))\n    random_indices = np.random.choice(len(X_scaled), sample_size, replace=False)\n    X_sample = X_scaled[random_indices]\n    \n    print(f\"Running MiniBatchKMeans + silhouette on {sample_size} samples\")\n    \n    start_time = time.time()\n    for k in k_range:\n        # Use MiniBatchKMeans, which is much faster than regular KMeans\n        minibatch_kmeans = MiniBatchKMeans(n_clusters=k, random_state=42, \n                                           batch_size=1024, n_init='auto')\n        # Fit on entire dataset\n        minibatch_kmeans.fit(X_scaled)\n        # Predict on the sample\n        cluster_labels = minibatch_kmeans.predict(X_sample)\n        # Calculate silhouette score on the sample\n        silhouette_avg.append(silhouette_score(X_sample, cluster_labels))\n        print(f\"k={k} completed in {time.time() - start_time:.2f} seconds\")\n        start_time = time.time()\n    \n    return k_range, silhouette_avg","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:57.436734Z","iopub.execute_input":"2025-03-03T00:35:57.437016Z","iopub.status.idle":"2025-03-03T00:35:57.443422Z","shell.execute_reply.started":"2025-03-03T00:35:57.436962Z","shell.execute_reply":"2025-03-03T00:35:57.442410Z"}},"outputs":[],"execution_count":14},{"cell_type":"code","source":"k_range, silhouette_avg = silhouette_analysis_with_minibatch()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:35:57.444362Z","iopub.execute_input":"2025-03-03T00:35:57.444629Z","iopub.status.idle":"2025-03-03T00:36:09.351085Z","shell.execute_reply.started":"2025-03-03T00:35:57.444605Z","shell.execute_reply":"2025-03-03T00:36:09.350201Z"}},"outputs":[{"name":"stdout","text":"Running MiniBatchKMeans + silhouette on 9917 samples\nk=2 completed in 1.58 seconds\nk=3 completed in 1.33 seconds\nk=4 completed in 1.31 seconds\nk=5 completed in 1.27 seconds\nk=6 completed in 1.27 seconds\nk=7 completed in 1.25 seconds\nk=8 completed in 1.28 seconds\nk=9 completed in 1.31 seconds\nk=10 completed in 1.27 seconds\n","output_type":"stream"}],"execution_count":15},{"cell_type":"code","source":"# Plot the results\nplt.figure(figsize=(10, 6))\nplt.plot(k_range, silhouette_avg, 'o-')\nplt.xlabel('Number of clusters (k)')\nplt.ylabel('Average Silhouette Score')\nplt.title('Silhouette Analysis for Optimal k')\nplt.grid(True)\nplt.savefig('silhouette_analysis.png')\nplt.close()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:09.351944Z","iopub.execute_input":"2025-03-03T00:36:09.352319Z","iopub.status.idle":"2025-03-03T00:36:09.547939Z","shell.execute_reply.started":"2025-03-03T00:36:09.352284Z","shell.execute_reply":"2025-03-03T00:36:09.546904Z"}},"outputs":[],"execution_count":16},{"cell_type":"code","source":"best_k_indices = np.argsort(silhouette_avg)[-3:]\nbest_k_values = [k_range[i] for i in best_k_indices]\nprint(f\"Best k values based on silhouette analysis: {best_k_values}\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:09.548864Z","iopub.execute_input":"2025-03-03T00:36:09.549232Z","iopub.status.idle":"2025-03-03T00:36:09.554349Z","shell.execute_reply.started":"2025-03-03T00:36:09.549198Z","shell.execute_reply":"2025-03-03T00:36:09.553349Z"}},"outputs":[{"name":"stdout","text":"Best k values based on silhouette analysis: [2, 4, 3]\n","output_type":"stream"}],"execution_count":17},{"cell_type":"code","source":"# Perform K-means clustering for the three best k values\nfor k in best_k_values:\n    kmeans = KMeans(n_clusters=k, random_state=42, n_init=10)\n    gdsc_data['kmeans_cluster'] = kmeans.fit_predict(X_scaled)\n    centroids = kmeans.cluster_centers_\n    \n    # Create a 2D visualization using the first two features\n    plt.figure(figsize=(12, 10))\n    \n    # Plot samples colored by LN_IC50\n    scatter = plt.scatter(X_scaled[:, 0], X_scaled[:, 1], c=y, cmap='viridis', \n                          alpha=0.6, edgecolor='k', s=50)\n    \n    # Plot centroids\n    plt.scatter(centroids[:, 0], centroids[:, 1], c='red', marker='X', s=200, \n                edgecolor='k', label='Centroids')\n    \n    plt.colorbar(scatter, label='LN_IC50')\n    plt.title(f'K-means Clustering (k={k})')\n    plt.xlabel(features[0])\n    plt.ylabel(features[1])\n    plt.legend()\n    plt.savefig(f'kmeans_k{k}.png')\n    plt.close()\n    \n    # Calculate and display distribution of target variable per cluster\n    cluster_stats = gdsc_data.groupby('kmeans_cluster')['LN_IC50'].agg(['mean', 'std', 'count'])\n    print(f\"\\nCluster statistics for k={k}:\")\n    print(cluster_stats)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:09.555293Z","iopub.execute_input":"2025-03-03T00:36:09.555570Z","iopub.status.idle":"2025-03-03T00:36:38.674643Z","shell.execute_reply.started":"2025-03-03T00:36:09.555522Z","shell.execute_reply":"2025-03-03T00:36:38.673659Z"}},"outputs":[{"name":"stdout","text":"\nCluster statistics for k=2:\n                    mean       std   count\nkmeans_cluster                            \n0               2.608734  1.645872  133777\n1               5.160238  1.246684   64565\n\nCluster statistics for k=4:\n                    mean       std  count\nkmeans_cluster                           \n0               2.827848  1.170628  30897\n1               1.030084  1.363876  40707\n2               3.903628  1.313326  88926\n3               5.440643  1.220716  37812\n\nCluster statistics for k=3:\n                    mean       std   count\nkmeans_cluster                            \n0               3.707740  1.281583  107600\n1               1.190558  1.352907   49101\n2               5.397299  1.239226   41641\n","output_type":"stream"}],"execution_count":18},{"cell_type":"code","source":"sample_size = min(5000, len(X_scaled))\nindices = np.random.choice(range(len(X_scaled)), sample_size, replace=False)\nX_sample = X_scaled[indices]\ny_sample = y.iloc[indices]","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:38.675640Z","iopub.execute_input":"2025-03-03T00:36:38.675894Z","iopub.status.idle":"2025-03-03T00:36:38.707356Z","shell.execute_reply.started":"2025-03-03T00:36:38.675873Z","shell.execute_reply":"2025-03-03T00:36:38.706483Z"}},"outputs":[],"execution_count":19},{"cell_type":"code","source":"# Calculate cosine similarity on the sample\nprint(\"Computing distance matrix...\")\ncosine_distances = pdist(X_sample, metric='cosine')","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:38.708258Z","iopub.execute_input":"2025-03-03T00:36:38.708607Z","iopub.status.idle":"2025-03-03T00:36:38.808182Z","shell.execute_reply.started":"2025-03-03T00:36:38.708569Z","shell.execute_reply":"2025-03-03T00:36:38.807139Z"}},"outputs":[{"name":"stdout","text":"Computing distance matrix...\n","output_type":"stream"}],"execution_count":20},{"cell_type":"code","source":"# Perform hierarchical clustering\nprint(\"Computing linkage...\")\nZ = linkage(cosine_distances, method='average')\n\n# Plot dendrogram\nplt.figure(figsize=(16, 10))\nplt.title('Hierarchical Clustering Dendrogram')\nplt.xlabel('Sample index')\nplt.ylabel('Distance (Cosine)')\n\n# Make dendrogram more readable by truncating\ndendrogram(Z, truncate_mode='lastp', p=30, leaf_font_size=10, \n          color_threshold=0.7 * max(Z[:, 2]))\n\nplt.savefig('hierarchical_clustering_dendrogram.png')\nplt.close()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:38.809097Z","iopub.execute_input":"2025-03-03T00:36:38.809329Z","iopub.status.idle":"2025-03-03T00:36:39.951433Z","shell.execute_reply.started":"2025-03-03T00:36:38.809308Z","shell.execute_reply":"2025-03-03T00:36:39.950480Z"}},"outputs":[{"name":"stdout","text":"Computing linkage...\n","output_type":"stream"}],"execution_count":21},{"cell_type":"code","source":"# Cut the dendrogram to get clusters (using same number as best k from k-means)\nfor k in best_k_values:\n    hierarchical_clusters = fcluster(Z, k, criterion='maxclust')\n    \n    # Plot 2D visualization of hierarchical clusters\n    plt.figure(figsize=(12, 10))\n    plt.scatter(X_sample[:, 0], X_sample[:, 1], c=hierarchical_clusters, \n                cmap='viridis', alpha=0.6, edgecolor='k', s=50)\n    plt.title(f'Hierarchical Clustering (k={k})')\n    plt.xlabel(features[0])\n    plt.ylabel(features[1])\n    plt.colorbar(label='Cluster')\n    plt.savefig(f'hierarchical_k{k}.png')\n    plt.close()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:39.952456Z","iopub.execute_input":"2025-03-03T00:36:39.952736Z","iopub.status.idle":"2025-03-03T00:36:41.499670Z","shell.execute_reply.started":"2025-03-03T00:36:39.952712Z","shell.execute_reply":"2025-03-03T00:36:41.498703Z"}},"outputs":[],"execution_count":22},{"cell_type":"code","source":"# DBSCAN Clustering\nfrom sklearn.neighbors import NearestNeighbors\n\nsample_size_knn = min(10000, len(X_scaled))\nrandom_indices_knn = np.random.choice(len(X_scaled), sample_size_knn, replace=False)\nX_sample_knn = X_scaled[random_indices_knn]\nprint(\"Performing DBSCAN Clustering...\")\n\nsample_size_knn = min(10000, len(X_scaled))\nrandom_indices_knn = np.random.choice(len(X_scaled), sample_size_knn, replace=False)\nX_sample_knn = X_scaled[random_indices_knn]","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:41.500593Z","iopub.execute_input":"2025-03-03T00:36:41.500841Z","iopub.status.idle":"2025-03-03T00:36:41.519988Z","shell.execute_reply.started":"2025-03-03T00:36:41.500819Z","shell.execute_reply":"2025-03-03T00:36:41.518968Z"}},"outputs":[{"name":"stdout","text":"Performing DBSCAN Clustering...\n","output_type":"stream"}],"execution_count":23},{"cell_type":"code","source":"# Calculate distances to 5th nearest neighbor\nneighbors = NearestNeighbors(n_neighbors=5)\nneighbors.fit(X_sample_knn)\ndistances, _ = neighbors.kneighbors(X_sample_knn)\ndistances = np.sort(distances[:, 4]) ","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:41.520966Z","iopub.execute_input":"2025-03-03T00:36:41.521241Z","iopub.status.idle":"2025-03-03T00:36:41.581984Z","shell.execute_reply.started":"2025-03-03T00:36:41.521217Z","shell.execute_reply":"2025-03-03T00:36:41.581078Z"}},"outputs":[],"execution_count":24},{"cell_type":"code","source":"# Plotting k-distance graph\nplt.figure(figsize=(10, 6))\nplt.plot(distances)\nplt.ylabel('Distance to 5th nearest neighbor')\nplt.xlabel('Points (sorted)')\nplt.title('K-distance Graph for DBSCAN Epsilon Parameter Selection')\nplt.grid(True)\nplt.savefig('dbscan_kdistance.png')\nplt.close()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:41.582814Z","iopub.execute_input":"2025-03-03T00:36:41.583068Z","iopub.status.idle":"2025-03-03T00:36:41.763446Z","shell.execute_reply.started":"2025-03-03T00:36:41.583047Z","shell.execute_reply":"2025-03-03T00:36:41.762626Z"}},"outputs":[],"execution_count":25},{"cell_type":"code","source":"# Find the \"elbow\" point\nfrom kneed import KneeLocator\nkneedle = KneeLocator(range(len(distances)), distances, S=1.0, \n                     curve='convex', direction='increasing')\neps = distances[kneedle.knee]\nprint(f\"Estimated epsilon: {eps}\")\n\n# Run DBSCAN with optimized parameters\ndbscan = DBSCAN(eps=eps, min_samples=5)\ndbscan_sample_size = min(50000, len(X_scaled))\nrandom_indices_dbscan = np.random.choice(len(X_scaled), dbscan_sample_size, replace=False)\nX_sample_dbscan = X_scaled[random_indices_dbscan]\ny_sample_dbscan = y.iloc[random_indices_dbscan]\n\ndbscan_labels = dbscan.fit_predict(X_sample_dbscan)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:41.767242Z","iopub.execute_input":"2025-03-03T00:36:41.767494Z","iopub.status.idle":"2025-03-03T00:36:48.267236Z","shell.execute_reply.started":"2025-03-03T00:36:41.767473Z","shell.execute_reply":"2025-03-03T00:36:48.266301Z"}},"outputs":[{"name":"stdout","text":"Estimated epsilon: 0.5656658525046282\n","output_type":"stream"}],"execution_count":26},{"cell_type":"code","source":"# Plot DBSCAN results\nplt.figure(figsize=(12, 10))\nunique_labels = np.unique(dbscan_labels)\ncolors = plt.cm.rainbow(np.linspace(0, 1, len(unique_labels)))\n\nfor label, col in zip(unique_labels, colors):\n    if label == -1:  # Noise points\n        col = 'black'\n    mask = dbscan_labels == label\n    plt.scatter(X_sample_dbscan[mask, 0], X_sample_dbscan[mask, 1], \n               c=[col], label=f'Cluster {label}' if label != -1 else 'Noise',\n               s=50, alpha=0.6, edgecolor='k')\n\nplt.title('DBSCAN Clustering')\nplt.xlabel(features[0])\nplt.ylabel(features[1])\nplt.legend()\nplt.savefig('dbscan_clustering.png')\nplt.close()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:48.268501Z","iopub.execute_input":"2025-03-03T00:36:48.269054Z","iopub.status.idle":"2025-03-03T00:36:49.981070Z","shell.execute_reply.started":"2025-03-03T00:36:48.269018Z","shell.execute_reply":"2025-03-03T00:36:49.980097Z"}},"outputs":[],"execution_count":27},{"cell_type":"code","source":"# Counting number of points in each cluster and noise\ncluster_counts = np.bincount(dbscan_labels[dbscan_labels >= 0])\nnoise_count = np.sum(dbscan_labels == -1)\nprint(f\"Number of clusters found by DBSCAN: {len(cluster_counts)}\")\nprint(f\"Number of noise points: {noise_count} ({noise_count/len(dbscan_labels)*100:.2f}%)\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:49.982034Z","iopub.execute_input":"2025-03-03T00:36:49.982378Z","iopub.status.idle":"2025-03-03T00:36:49.988928Z","shell.execute_reply.started":"2025-03-03T00:36:49.982342Z","shell.execute_reply":"2025-03-03T00:36:49.988155Z"}},"outputs":[{"name":"stdout","text":"Number of clusters found by DBSCAN: 7\nNumber of noise points: 25 (0.05%)\n","output_type":"stream"}],"execution_count":28},{"cell_type":"code","source":"# Calculate statistics for each DBSCAN cluster\ndbscan_df = pd.DataFrame({\n    'dbscan_cluster': dbscan_labels,\n    'LN_IC50': y_sample_dbscan.values\n})\n\nfor label in unique_labels:\n    cluster_data = dbscan_df[dbscan_df['dbscan_cluster'] == label]\n    label_name = 'Noise' if label == -1 else f'Cluster {label}'\n    print(f\"\\n{label_name} statistics:\")\n    print(f\"Count: {len(cluster_data)}\")\n    if len(cluster_data) > 0:\n        print(f\"Mean LN_IC50: {cluster_data['LN_IC50'].mean():.4f}\")\n        print(f\"Std LN_IC50: {cluster_data['LN_IC50'].std():.4f}\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:49.989825Z","iopub.execute_input":"2025-03-03T00:36:49.990108Z","iopub.status.idle":"2025-03-03T00:36:50.019584Z","shell.execute_reply.started":"2025-03-03T00:36:49.990084Z","shell.execute_reply":"2025-03-03T00:36:50.018533Z"}},"outputs":[{"name":"stdout","text":"\nNoise statistics:\nCount: 25\nMean LN_IC50: 1.7531\nStd LN_IC50: 3.4260\n\nCluster 0 statistics:\nCount: 49735\nMean LN_IC50: 3.4239\nStd LN_IC50: 1.9020\n\nCluster 1 statistics:\nCount: 37\nMean LN_IC50: -2.5078\nStd LN_IC50: 0.4973\n\nCluster 2 statistics:\nCount: 164\nMean LN_IC50: 8.8495\nStd LN_IC50: 0.4491\n\nCluster 3 statistics:\nCount: 8\nMean LN_IC50: 0.1707\nStd LN_IC50: 0.2887\n\nCluster 4 statistics:\nCount: 12\nMean LN_IC50: -2.6728\nStd LN_IC50: 0.3557\n\nCluster 5 statistics:\nCount: 14\nMean LN_IC50: 7.6511\nStd LN_IC50: 0.3752\n\nCluster 6 statistics:\nCount: 5\nMean LN_IC50: -3.1678\nStd LN_IC50: 0.0685\n","output_type":"stream"}],"execution_count":29},{"cell_type":"code","source":"# Compare clustering results\nprint(\"\\nComparison of Clustering Methods:\")\nprint(\"1. K-means identified spherical clusters based on distance to centroids\")\nprint(\"2. Hierarchical clustering grouped samples based on cosine similarity\")\nprint(f\"3. DBSCAN found {len(cluster_counts)} clusters and {noise_count} noise points based on density\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:50.020767Z","iopub.execute_input":"2025-03-03T00:36:50.021231Z","iopub.status.idle":"2025-03-03T00:36:50.026748Z","shell.execute_reply.started":"2025-03-03T00:36:50.021196Z","shell.execute_reply":"2025-03-03T00:36:50.025957Z"}},"outputs":[{"name":"stdout","text":"\nComparison of Clustering Methods:\n1. K-means identified spherical clusters based on distance to centroids\n2. Hierarchical clustering grouped samples based on cosine similarity\n3. DBSCAN found 7 clusters and 25 noise points based on density\n","output_type":"stream"}],"execution_count":30},{"cell_type":"code","source":"# Create a mapping from the DBSCAN sample back to the original indices\nsample_indices_map = {i: idx for i, idx in enumerate(random_indices_dbscan)}\n\n# Initialize an array for all data points (default to -1 for noise)\nall_dbscan_labels = np.full(len(gdsc_data), -1)\n\n# Fill in the labels for the sampled points\nfor i, label in enumerate(dbscan_labels):\n    original_idx = random_indices_dbscan[i]\n    all_dbscan_labels[original_idx] = label\n\n# Add DBSCAN cluster labels to the main dataframe\ngdsc_data['dbscan_cluster'] = all_dbscan_labels\n\n# Now calculate variance explained by each clustering method\nkmeans_variance = gdsc_data.groupby('kmeans_cluster')['LN_IC50'].var().sum() / gdsc_data['LN_IC50'].var()\ndbscan_variance = gdsc_data[gdsc_data['dbscan_cluster'] >= 0].groupby('dbscan_cluster')['LN_IC50'].var().sum() / gdsc_data['LN_IC50'].var()\n\nprint(\"\\nVariance Explained in Drug Sensitivity (LN_IC50):\")\nprint(f\"K-means: {(1 - kmeans_variance) * 100:.2f}%\")\nprint(f\"DBSCAN: {(1 - dbscan_variance) * 100:.2f}%\")\n\n# For hierarchical clustering, create a similar mapping\n# First initialize an array for all data points\nall_hierarchical_labels = np.full(len(gdsc_data), -1)\n\n# Fill in the labels for the sampled points\nfor i, label in enumerate(hierarchical_clusters):\n    original_idx = indices[i]\n    all_hierarchical_labels[original_idx] = label\n\n# Add hierarchical cluster labels to the main dataframe\ngdsc_data['hierarchical_cluster'] = all_hierarchical_labels\n\n# Calculate variance explained by hierarchical clustering\nhierarchical_variance = gdsc_data[gdsc_data['hierarchical_cluster'] >= 0].groupby('hierarchical_cluster')['LN_IC50'].var().sum() / gdsc_data['LN_IC50'].var()\nprint(f\"Hierarchical: {(1 - hierarchical_variance) * 100:.2f}%\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-03-03T00:36:50.027777Z","iopub.execute_input":"2025-03-03T00:36:50.028367Z","iopub.status.idle":"2025-03-03T00:36:50.107468Z","shell.execute_reply.started":"2025-03-03T00:36:50.028331Z","shell.execute_reply":"2025-03-03T00:36:50.106425Z"}},"outputs":[{"name":"stdout","text":"\nVariance Explained in Drug Sensitivity (LN_IC50):\nK-means: -33.12%\nDBSCAN: -17.53%\nHierarchical: -17.65%\n","output_type":"stream"}],"execution_count":31}]}