{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.10.12","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kaggle":{"accelerator":"none","dataSources":[{"sourceId":10885300,"sourceType":"datasetVersion","datasetId":6764139}],"dockerImageVersionId":30918,"isInternetEnabled":true,"language":"python","sourceType":"notebook","isGpuEnabled":false}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:22.134379Z","iopub.execute_input":"2025-04-07T04:03:22.135163Z","iopub.status.idle":"2025-04-07T04:03:22.567751Z","shell.execute_reply.started":"2025-04-07T04:03:22.135101Z","shell.execute_reply":"2025-04-07T04:03:22.565986Z"}},"outputs":[{"name":"stdout","text":"/kaggle/input/cleaned-genomics/cleaned_data_no_outliers.csv\n","output_type":"stream"}],"execution_count":1},{"cell_type":"code","source":"import pandas as pd\nimport numpy as np\nimport matplotlib.pyplot as plt\nimport seaborn as sns\nfrom sklearn.model_selection import train_test_split, GridSearchCV\nfrom sklearn.preprocessing import OneHotEncoder, StandardScaler\nfrom sklearn.tree import DecisionTreeClassifier, plot_tree, export_graphviz\nfrom sklearn.metrics import confusion_matrix, classification_report, accuracy_score\nfrom sklearn.compose import ColumnTransformer\nfrom sklearn.pipeline import Pipeline\nimport graphviz\nfrom IPython.display import display","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:22.569094Z","iopub.execute_input":"2025-04-07T04:03:22.569579Z","iopub.status.idle":"2025-04-07T04:03:23.272848Z","shell.execute_reply.started":"2025-04-07T04:03:22.569537Z","shell.execute_reply":"2025-04-07T04:03:23.272084Z"}},"outputs":[],"execution_count":2},{"cell_type":"code","source":"# Load the dataset\ndata = pd.read_csv('/kaggle/input/cleaned-genomics/cleaned_data_no_outliers.csv')","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:23.274277Z","iopub.execute_input":"2025-04-07T04:03:23.274656Z","iopub.status.idle":"2025-04-07T04:03:23.696849Z","shell.execute_reply.started":"2025-04-07T04:03:23.274636Z","shell.execute_reply":"2025-04-07T04:03:23.693872Z"}},"outputs":[],"execution_count":3},{"cell_type":"code","source":"output_dir = \"/kaggle/working/\"","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:23.698547Z","iopub.execute_input":"2025-04-07T04:03:23.701739Z","iopub.status.idle":"2025-04-07T04:03:23.709126Z","shell.execute_reply.started":"2025-04-07T04:03:23.701692Z","shell.execute_reply":"2025-04-07T04:03:23.707987Z"}},"outputs":[],"execution_count":4},{"cell_type":"code","source":"data.head().T","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:23.713061Z","iopub.execute_input":"2025-04-07T04:03:23.713372Z","iopub.status.idle":"2025-04-07T04:03:23.762684Z","shell.execute_reply.started":"2025-04-07T04:03:23.713343Z","shell.execute_reply":"2025-04-07T04:03:23.760396Z"}},"outputs":[{"execution_count":5,"output_type":"execute_result","data":{"text/plain":"                                                       0                1  \\\nCOSMIC_ID                                         683667           687448   \nCELL_LINE_NAME                                    PFSK-1         COLO-829   \nTCGA_DESC                                             MB             SKCM   \nDRUG_ID                                             1003             1003   \nDRUG_NAME                                   Camptothecin     Camptothecin   \nLN_IC50                                        -1.463887        -1.235034   \nAUC                                              0.93022         0.867348   \nZ_SCORE                                         0.433123         0.557727   \nGDSC Tissue descriptor 1                  nervous_system             skin   \nGDSC Tissue descriptor 2                 medulloblastoma         melanoma   \nCancer Type (matching TCGA label)                     MB             SKCM   \nMicrosatellite instability Status (MSI)        MSS/MSI-L        MSS/MSI-L   \nScreen Medium                                          R                R   \nGrowth Properties                               Adherent         Adherent   \nCNA                                                    1                1   \nGene Expression                                        1                1   \nMethylation                                            1                1   \nTARGET                                              TOP1             TOP1   \nTARGET_PATHWAY                           DNA replication  DNA replication   \n\n                                                       2                  3  \\\nCOSMIC_ID                                         687452             687455   \nCELL_LINE_NAME                                      5637                RT4   \nTCGA_DESC                                           BLCA               BLCA   \nDRUG_ID                                             1003               1003   \nDRUG_NAME                                   Camptothecin       Camptothecin   \nLN_IC50                                        -2.632632          -2.963191   \nAUC                                             0.834067           0.821438   \nZ_SCORE                                        -0.203221            -0.3832   \nGDSC Tissue descriptor 1                         Unknown  urogenital_system   \nGDSC Tissue descriptor 2                         Unknown            Bladder   \nCancer Type (matching TCGA label)                Unknown               BLCA   \nMicrosatellite instability Status (MSI)        MSS/MSI-L          MSS/MSI-L   \nScreen Medium                                          R              D/F12   \nGrowth Properties                               Adherent           Adherent   \nCNA                                                    1                  1   \nGene Expression                                        1                  1   \nMethylation                                            1                  1   \nTARGET                                              TOP1               TOP1   \nTARGET_PATHWAY                           DNA replication    DNA replication   \n\n                                                         4  \nCOSMIC_ID                                           687457  \nCELL_LINE_NAME                                       SW780  \nTCGA_DESC                                             BLCA  \nDRUG_ID                                               1003  \nDRUG_NAME                                     Camptothecin  \nLN_IC50                                          -1.449138  \nAUC                                                0.90505  \nZ_SCORE                                           0.441154  \nGDSC Tissue descriptor 1                 urogenital_system  \nGDSC Tissue descriptor 2                           Bladder  \nCancer Type (matching TCGA label)                     BLCA  \nMicrosatellite instability Status (MSI)          MSS/MSI-L  \nScreen Medium                                        D/F12  \nGrowth Properties                                 Adherent  \nCNA                                                      1  \nGene Expression                                          1  \nMethylation                                              1  \nTARGET                                                TOP1  \nTARGET_PATHWAY                             DNA replication  ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>0</th>\n      <th>1</th>\n      <th>2</th>\n      <th>3</th>\n      <th>4</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>COSMIC_ID</th>\n      <td>683667</td>\n      <td>687448</td>\n      <td>687452</td>\n      <td>687455</td>\n      <td>687457</td>\n    </tr>\n    <tr>\n      <th>CELL_LINE_NAME</th>\n      <td>PFSK-1</td>\n      <td>COLO-829</td>\n      <td>5637</td>\n      <td>RT4</td>\n      <td>SW780</td>\n    </tr>\n    <tr>\n      <th>TCGA_DESC</th>\n      <td>MB</td>\n      <td>SKCM</td>\n      <td>BLCA</td>\n      <td>BLCA</td>\n      <td>BLCA</td>\n    </tr>\n    <tr>\n      <th>DRUG_ID</th>\n      <td>1003</td>\n      <td>1003</td>\n      <td>1003</td>\n      <td>1003</td>\n      <td>1003</td>\n    </tr>\n    <tr>\n      <th>DRUG_NAME</th>\n      <td>Camptothecin</td>\n      <td>Camptothecin</td>\n      <td>Camptothecin</td>\n      <td>Camptothecin</td>\n      <td>Camptothecin</td>\n    </tr>\n    <tr>\n      <th>LN_IC50</th>\n      <td>-1.463887</td>\n      <td>-1.235034</td>\n      <td>-2.632632</td>\n      <td>-2.963191</td>\n      <td>-1.449138</td>\n    </tr>\n    <tr>\n      <th>AUC</th>\n      <td>0.93022</td>\n      <td>0.867348</td>\n      <td>0.834067</td>\n      <td>0.821438</td>\n      <td>0.90505</td>\n    </tr>\n    <tr>\n      <th>Z_SCORE</th>\n      <td>0.433123</td>\n      <td>0.557727</td>\n      <td>-0.203221</td>\n      <td>-0.3832</td>\n      <td>0.441154</td>\n    </tr>\n    <tr>\n      <th>GDSC Tissue descriptor 1</th>\n      <td>nervous_system</td>\n      <td>skin</td>\n      <td>Unknown</td>\n      <td>urogenital_system</td>\n      <td>urogenital_system</td>\n    </tr>\n    <tr>\n      <th>GDSC Tissue descriptor 2</th>\n      <td>medulloblastoma</td>\n      <td>melanoma</td>\n      <td>Unknown</td>\n      <td>Bladder</td>\n      <td>Bladder</td>\n    </tr>\n    <tr>\n      <th>Cancer Type (matching TCGA label)</th>\n      <td>MB</td>\n      <td>SKCM</td>\n      <td>Unknown</td>\n      <td>BLCA</td>\n      <td>BLCA</td>\n    </tr>\n    <tr>\n      <th>Microsatellite instability Status (MSI)</th>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n    </tr>\n    <tr>\n      <th>Screen Medium</th>\n      <td>R</td>\n      <td>R</td>\n      <td>R</td>\n      <td>D/F12</td>\n      <td>D/F12</td>\n    </tr>\n    <tr>\n      <th>Growth Properties</th>\n      <td>Adherent</td>\n      <td>Adherent</td>\n      <td>Adherent</td>\n      <td>Adherent</td>\n      <td>Adherent</td>\n    </tr>\n    <tr>\n      <th>CNA</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>Gene Expression</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>Methylation</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>TARGET</th>\n      <td>TOP1</td>\n      <td>TOP1</td>\n      <td>TOP1</td>\n      <td>TOP1</td>\n      <td>TOP1</td>\n    </tr>\n    <tr>\n      <th>TARGET_PATHWAY</th>\n      <td>DNA replication</td>\n      <td>DNA replication</td>\n      <td>DNA replication</td>\n      <td>DNA replication</td>\n      <td>DNA replication</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}],"execution_count":5},{"cell_type":"code","source":"# Create a binary target variable based on LN_IC50\n# Lower values indicate higher sensitivity\ndata['Sensitivity'] = np.where(data['LN_IC50'] <= data['LN_IC50'].median(), 'Sensitive', 'Resistant')\n\n# Select features for modeling\nfeatures = ['TCGA_DESC', 'GDSC Tissue descriptor 1', 'GDSC Tissue descriptor 2', \n           'Cancer Type (matching TCGA label)', 'Microsatellite instability Status (MSI)',\n           'Growth Properties', 'CNA', 'Gene Expression', 'Methylation', \n           'TARGET', 'TARGET_PATHWAY']\n\nX = data[features]\ny = data['Sensitivity']","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:23.763228Z","iopub.execute_input":"2025-04-07T04:03:23.763493Z","iopub.status.idle":"2025-04-07T04:03:23.820099Z","shell.execute_reply.started":"2025-04-07T04:03:23.763456Z","shell.execute_reply":"2025-04-07T04:03:23.819274Z"}},"outputs":[],"execution_count":6},{"cell_type":"code","source":"# Split data into training and testing sets (disjoint)\nX_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)\n\n# Preprocessing for categorical and numerical features\ncategorical_features = ['TCGA_DESC', 'GDSC Tissue descriptor 1', 'GDSC Tissue descriptor 2', \n                        'Cancer Type (matching TCGA label)', 'Microsatellite instability Status (MSI)',\n                        'Growth Properties', 'TARGET', 'TARGET_PATHWAY']\nnumerical_features = ['CNA', 'Gene Expression', 'Methylation']\n\n# Create preprocessing steps\npreprocessor = ColumnTransformer(\n    transformers=[\n        ('num', StandardScaler(), numerical_features),\n        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_features)\n    ])","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:23.820873Z","iopub.execute_input":"2025-04-07T04:03:23.821110Z","iopub.status.idle":"2025-04-07T04:03:24.066873Z","shell.execute_reply.started":"2025-04-07T04:03:23.821086Z","shell.execute_reply":"2025-04-07T04:03:24.065696Z"}},"outputs":[],"execution_count":7},{"cell_type":"code","source":"# Fit the preprocessor on training data to get transformed feature names\nX_train_preprocessed = preprocessor.fit_transform(X_train)\n\n# Get feature names after preprocessing\nohe = preprocessor.named_transformers_['cat']\ncat_feature_names = ohe.get_feature_names_out(categorical_features)\nprocessed_feature_names = np.append(numerical_features, cat_feature_names)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:24.067720Z","iopub.execute_input":"2025-04-07T04:03:24.067980Z","iopub.status.idle":"2025-04-07T04:03:24.341411Z","shell.execute_reply.started":"2025-04-07T04:03:24.067956Z","shell.execute_reply":"2025-04-07T04:03:24.340211Z"}},"outputs":[],"execution_count":8},{"cell_type":"code","source":"X_train.head().T","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:24.343373Z","iopub.execute_input":"2025-04-07T04:03:24.343701Z","iopub.status.idle":"2025-04-07T04:03:24.353206Z","shell.execute_reply.started":"2025-04-07T04:03:24.343675Z","shell.execute_reply":"2025-04-07T04:03:24.352244Z"}},"outputs":[{"execution_count":9,"output_type":"execute_result","data":{"text/plain":"                                                   173490             166374  \\\nTCGA_DESC                                            HNSC               BLCA   \nGDSC Tissue descriptor 1                   aero_dig_tract  urogenital_system   \nGDSC Tissue descriptor 2                    head and neck            Bladder   \nCancer Type (matching TCGA label)                    HNSC               BLCA   \nMicrosatellite instability Status (MSI)         MSS/MSI-L          MSS/MSI-L   \nGrowth Properties                                Adherent           Adherent   \nCNA                                                     1                  1   \nGene Expression                                         1                  1   \nMethylation                                             1                  1   \nTARGET                                               TERT            Unknown   \nTARGET_PATHWAY                           Genome integrity       Unclassified   \n\n                                                  104586  \\\nTCGA_DESC                                            ALL   \nGDSC Tissue descriptor 1                        leukemia   \nGDSC Tissue descriptor 2                 T_cell_leukemia   \nCancer Type (matching TCGA label)                    ALL   \nMicrosatellite instability Status (MSI)            MSI-H   \nGrowth Properties                             Suspension   \nCNA                                                    1   \nGene Expression                                        1   \nMethylation                                            1   \nTARGET                                              JAK1   \nTARGET_PATHWAY                            Other, kinases   \n\n                                                                181915  \\\nTCGA_DESC                                                         LUAD   \nGDSC Tissue descriptor 1                                    lung_NSCLC   \nGDSC Tissue descriptor 2                     lung_NSCLC_adenocarcinoma   \nCancer Type (matching TCGA label)                                 LUAD   \nMicrosatellite instability Status (MSI)                      MSS/MSI-L   \nGrowth Properties                                             Adherent   \nCNA                                                                  1   \nGene Expression                                                      1   \nMethylation                                                          1   \nTARGET                                                           PMRT5   \nTARGET_PATHWAY                           Chromatin histone methylation   \n\n                                                                83290   \nTCGA_DESC                                                         LUAD  \nGDSC Tissue descriptor 1                                    lung_NSCLC  \nGDSC Tissue descriptor 2                     lung_NSCLC_adenocarcinoma  \nCancer Type (matching TCGA label)                                 LUAD  \nMicrosatellite instability Status (MSI)                      MSS/MSI-L  \nGrowth Properties                                             Adherent  \nCNA                                                                  1  \nGene Expression                                                      1  \nMethylation                                                          1  \nTARGET                                                    BAZ2A, BAZ2B  \nTARGET_PATHWAY                           Chromatin histone acetylation  ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>173490</th>\n      <th>166374</th>\n      <th>104586</th>\n      <th>181915</th>\n      <th>83290</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>TCGA_DESC</th>\n      <td>HNSC</td>\n      <td>BLCA</td>\n      <td>ALL</td>\n      <td>LUAD</td>\n      <td>LUAD</td>\n    </tr>\n    <tr>\n      <th>GDSC Tissue descriptor 1</th>\n      <td>aero_dig_tract</td>\n      <td>urogenital_system</td>\n      <td>leukemia</td>\n      <td>lung_NSCLC</td>\n      <td>lung_NSCLC</td>\n    </tr>\n    <tr>\n      <th>GDSC Tissue descriptor 2</th>\n      <td>head and neck</td>\n      <td>Bladder</td>\n      <td>T_cell_leukemia</td>\n      <td>lung_NSCLC_adenocarcinoma</td>\n      <td>lung_NSCLC_adenocarcinoma</td>\n    </tr>\n    <tr>\n      <th>Cancer Type (matching TCGA label)</th>\n      <td>HNSC</td>\n      <td>BLCA</td>\n      <td>ALL</td>\n      <td>LUAD</td>\n      <td>LUAD</td>\n    </tr>\n    <tr>\n      <th>Microsatellite instability Status (MSI)</th>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n      <td>MSI-H</td>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n    </tr>\n    <tr>\n      <th>Growth Properties</th>\n      <td>Adherent</td>\n      <td>Adherent</td>\n      <td>Suspension</td>\n      <td>Adherent</td>\n      <td>Adherent</td>\n    </tr>\n    <tr>\n      <th>CNA</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>Gene Expression</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>Methylation</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>TARGET</th>\n      <td>TERT</td>\n      <td>Unknown</td>\n      <td>JAK1</td>\n      <td>PMRT5</td>\n      <td>BAZ2A, BAZ2B</td>\n    </tr>\n    <tr>\n      <th>TARGET_PATHWAY</th>\n      <td>Genome integrity</td>\n      <td>Unclassified</td>\n      <td>Other, kinases</td>\n      <td>Chromatin histone methylation</td>\n      <td>Chromatin histone acetylation</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}],"execution_count":9},{"cell_type":"code","source":"X_test.head().T","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:24.354876Z","iopub.execute_input":"2025-04-07T04:03:24.355375Z","iopub.status.idle":"2025-04-07T04:03:24.384947Z","shell.execute_reply.started":"2025-04-07T04:03:24.355333Z","shell.execute_reply":"2025-04-07T04:03:24.383865Z"}},"outputs":[{"execution_count":10,"output_type":"execute_result","data":{"text/plain":"                                                        22635   \\\nTCGA_DESC                                                 HNSC   \nGDSC Tissue descriptor 1                        aero_dig_tract   \nGDSC Tissue descriptor 2                         head and neck   \nCancer Type (matching TCGA label)                         HNSC   \nMicrosatellite instability Status (MSI)              MSS/MSI-L   \nGrowth Properties                                     Adherent   \nCNA                                                          1   \nGene Expression                                              1   \nMethylation                                                  1   \nTARGET                                                     JNK   \nTARGET_PATHWAY                           JNK and p38 signaling   \n\n                                                                170739  \\\nTCGA_DESC                                                         SKCM   \nGDSC Tissue descriptor 1                                          skin   \nGDSC Tissue descriptor 2                                      melanoma   \nCancer Type (matching TCGA label)                                 SKCM   \nMicrosatellite instability Status (MSI)                      MSS/MSI-L   \nGrowth Properties                                             Adherent   \nCNA                                                                  1   \nGene Expression                                                      1   \nMethylation                                                          1   \nTARGET                                                        G9A, GLP   \nTARGET_PATHWAY                           Chromatin histone methylation   \n\n                                                 183345                163720  \\\nTCGA_DESC                                          HNSC                COREAD   \nGDSC Tissue descriptor 1                 aero_dig_tract       large_intestine   \nGDSC Tissue descriptor 2                  head and neck       large_intestine   \nCancer Type (matching TCGA label)                  HNSC             COAD/READ   \nMicrosatellite instability Status (MSI)       MSS/MSI-L             MSS/MSI-L   \nGrowth Properties                              Adherent              Adherent   \nCNA                                                   1                     1   \nGene Expression                                       1                     1   \nMethylation                                           1                     1   \nTARGET                                          Unknown                  MCL1   \nTARGET_PATHWAY                             Unclassified  Apoptosis regulation   \n\n                                                                            144318  \nTCGA_DESC                                                             UNCLASSIFIED  \nGDSC Tissue descriptor 1                                                  leukemia  \nGDSC Tissue descriptor 2                                           B_cell_leukemia  \nCancer Type (matching TCGA label)                                          Unknown  \nMicrosatellite instability Status (MSI)                                      MSI-H  \nGrowth Properties                                                       Suspension  \nCNA                                                                              1  \nGene Expression                                                                  1  \nMethylation                                                                      1  \nTARGET                                   PI3Kalpha, PI3Kdelta, PI3Kbeta, PI3Kgamma  \nTARGET_PATHWAY                                                 PI3K/MTOR signaling  ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>22635</th>\n      <th>170739</th>\n      <th>183345</th>\n      <th>163720</th>\n      <th>144318</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>TCGA_DESC</th>\n      <td>HNSC</td>\n      <td>SKCM</td>\n      <td>HNSC</td>\n      <td>COREAD</td>\n      <td>UNCLASSIFIED</td>\n    </tr>\n    <tr>\n      <th>GDSC Tissue descriptor 1</th>\n      <td>aero_dig_tract</td>\n      <td>skin</td>\n      <td>aero_dig_tract</td>\n      <td>large_intestine</td>\n      <td>leukemia</td>\n    </tr>\n    <tr>\n      <th>GDSC Tissue descriptor 2</th>\n      <td>head and neck</td>\n      <td>melanoma</td>\n      <td>head and neck</td>\n      <td>large_intestine</td>\n      <td>B_cell_leukemia</td>\n    </tr>\n    <tr>\n      <th>Cancer Type (matching TCGA label)</th>\n      <td>HNSC</td>\n      <td>SKCM</td>\n      <td>HNSC</td>\n      <td>COAD/READ</td>\n      <td>Unknown</td>\n    </tr>\n    <tr>\n      <th>Microsatellite instability Status (MSI)</th>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n      <td>MSS/MSI-L</td>\n      <td>MSI-H</td>\n    </tr>\n    <tr>\n      <th>Growth Properties</th>\n      <td>Adherent</td>\n      <td>Adherent</td>\n      <td>Adherent</td>\n      <td>Adherent</td>\n      <td>Suspension</td>\n    </tr>\n    <tr>\n      <th>CNA</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>Gene Expression</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>Methylation</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>TARGET</th>\n      <td>JNK</td>\n      <td>G9A, GLP</td>\n      <td>Unknown</td>\n      <td>MCL1</td>\n      <td>PI3Kalpha, PI3Kdelta, PI3Kbeta, PI3Kgamma</td>\n    </tr>\n    <tr>\n      <th>TARGET_PATHWAY</th>\n      <td>JNK and p38 signaling</td>\n      <td>Chromatin histone methylation</td>\n      <td>Unclassified</td>\n      <td>Apoptosis regulation</td>\n      <td>PI3K/MTOR signaling</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}],"execution_count":10},{"cell_type":"code","source":"y_train","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:24.385825Z","iopub.execute_input":"2025-04-07T04:03:24.386070Z","iopub.status.idle":"2025-04-07T04:03:24.409047Z","shell.execute_reply.started":"2025-04-07T04:03:24.386047Z","shell.execute_reply":"2025-04-07T04:03:24.408110Z"}},"outputs":[{"execution_count":11,"output_type":"execute_result","data":{"text/plain":"173490    Resistant\n166374    Resistant\n104586    Resistant\n181915    Resistant\n83290     Resistant\n            ...    \n57986     Resistant\n6480      Sensitive\n129385    Sensitive\n56110     Sensitive\n107433    Resistant\nName: Sensitivity, Length: 158673, dtype: object"},"metadata":{}}],"execution_count":11},{"cell_type":"code","source":"y_test","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:24.410037Z","iopub.execute_input":"2025-04-07T04:03:24.410307Z","iopub.status.idle":"2025-04-07T04:03:24.432455Z","shell.execute_reply.started":"2025-04-07T04:03:24.410285Z","shell.execute_reply":"2025-04-07T04:03:24.431397Z"}},"outputs":[{"execution_count":12,"output_type":"execute_result","data":{"text/plain":"22635     Resistant\n170739    Resistant\n183345    Sensitive\n163720    Resistant\n144318    Sensitive\n            ...    \n3195      Resistant\n82721     Resistant\n36399     Sensitive\n140638    Resistant\n151919    Resistant\nName: Sensitivity, Length: 39669, dtype: object"},"metadata":{}}],"execution_count":12},{"cell_type":"code","source":"X_train.to_csv(f\"{output_dir}/DT_Xtrain.csv\", index=False)\nprint(f\"X_train dataset saved to {output_dir}/DT_Xtrain.csv\")\n\nX_test.to_csv(f\"{output_dir}/DT_Xtest.csv\", index=False)\nprint(f\"X_test dataset saved to {output_dir}/DT_Xtest.csv\")\n\ny_train.to_csv(f\"{output_dir}/DT_ytrain.csv\", index=False)\nprint(f\"y_train dataset saved to {output_dir}/DT_ytrain.csv\")\n\ny_test.to_csv(f\"{output_dir}/DT_ytest.csv\", index=False)\nprint(f\"y_test dataset saved to {output_dir}/DT_ytest.csv\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:24.433342Z","iopub.execute_input":"2025-04-07T04:03:24.433622Z","iopub.status.idle":"2025-04-07T04:03:25.094023Z","shell.execute_reply.started":"2025-04-07T04:03:24.433598Z","shell.execute_reply":"2025-04-07T04:03:25.092675Z"}},"outputs":[{"name":"stdout","text":"X_train dataset saved to /kaggle/working//DT_Xtrain.csv\nX_test dataset saved to /kaggle/working//DT_Xtest.csv\ny_train dataset saved to /kaggle/working//DT_ytrain.csv\ny_test dataset saved to /kaggle/working//DT_ytest.csv\n","output_type":"stream"}],"execution_count":13},{"cell_type":"code","source":"# Build three different decision tree models\n\n# Model 1: Standard Decision Tree\ndt1 = DecisionTreeClassifier(random_state=42)\ndt1.fit(X_train_preprocessed, y_train)\n\n# Model 2: Tree with entropy criterion and maximum depth constraint\ndt2 = DecisionTreeClassifier(random_state=42, criterion=\"entropy\", max_depth=15)\ndt2.fit(X_train_preprocessed, y_train)\n\n# Model 3: Tree with Gini criterion and minimum samples split constraint at a different depth\ndt3 = DecisionTreeClassifier(random_state=42, criterion=\"gini\", min_samples_split=20)\ndt3.fit(X_train_preprocessed, y_train)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:25.094892Z","iopub.execute_input":"2025-04-07T04:03:25.095145Z","iopub.status.idle":"2025-04-07T04:03:48.048784Z","shell.execute_reply.started":"2025-04-07T04:03:25.095121Z","shell.execute_reply":"2025-04-07T04:03:48.047598Z"}},"outputs":[{"execution_count":14,"output_type":"execute_result","data":{"text/plain":"DecisionTreeClassifier(min_samples_split=20, random_state=42)","text/html":"<style>#sk-container-id-1 {color: black;background-color: white;}#sk-container-id-1 pre{padding: 0;}#sk-container-id-1 div.sk-toggleable {background-color: white;}#sk-container-id-1 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-1 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-1 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-1 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-1 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-1 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-1 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-1 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-1 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-1 div.sk-item {position: relative;z-index: 1;}#sk-container-id-1 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-1 div.sk-item::before, #sk-container-id-1 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-1 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-1 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-1 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-1 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-1 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-1 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-1 div.sk-label-container {text-align: center;}#sk-container-id-1 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-1 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>DecisionTreeClassifier(min_samples_split=20, random_state=42)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">DecisionTreeClassifier</label><div class=\"sk-toggleable__content\"><pre>DecisionTreeClassifier(min_samples_split=20, random_state=42)</pre></div></div></div></div></div>"},"metadata":{}}],"execution_count":14},{"cell_type":"code","source":"# Transform test data for predictions\nX_test_preprocessed = preprocessor.transform(X_test)\n\n# Make predictions\ny_pred1 = dt1.predict(X_test_preprocessed)\ny_pred2 = dt2.predict(X_test_preprocessed)\ny_pred3 = dt3.predict(X_test_preprocessed)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:48.049618Z","iopub.execute_input":"2025-04-07T04:03:48.049828Z","iopub.status.idle":"2025-04-07T04:03:48.141685Z","shell.execute_reply.started":"2025-04-07T04:03:48.049805Z","shell.execute_reply":"2025-04-07T04:03:48.140168Z"}},"outputs":[],"execution_count":15},{"cell_type":"code","source":"# Evaluate models and save confusion matrices\ndef evaluate_model(y_true, y_pred, model_name):\n    accuracy = accuracy_score(y_true, y_pred)\n    conf_matrix = confusion_matrix(y_true, y_pred)\n    \n    print(f\"Model: {model_name}\")\n    print(f\"Accuracy: {accuracy:.4f}\")\n    print(\"Confusion Matrix:\")\n    print(conf_matrix)\n    print(\"Classification Report:\")\n    print(classification_report(y_true, y_pred))\n    print(\"\\n\")\n    \n    # Plot and save confusion matrix\n    plt.figure(figsize=(8, 6))\n    sns.heatmap(conf_matrix, annot=True, fmt='d', cmap='Blues',\n               xticklabels=['Resistant', 'Sensitive'],\n               yticklabels=['Resistant', 'Sensitive'])\n    plt.xlabel('Predicted')\n    plt.ylabel('Actual')\n    plt.title(f'Confusion Matrix - {model_name}')\n    plt.tight_layout()\n    \n    # Save to file\n    filename = f'/kaggle/working/confusion_matrix_{model_name.replace(\" \", \"_\")}.png'\n    plt.savefig(filename, dpi=300, bbox_inches='tight')\n    plt.close()\n    \n    return accuracy, conf_matrix\n\n# Evaluate all three models\naccuracy1, cm1 = evaluate_model(y_test, y_pred1, \"Standard Decision Tree\")\naccuracy2, cm2 = evaluate_model(y_test, y_pred2, \"Decision Tree with Max Depth 15 and criterion as Entropy\")\naccuracy3, cm3 = evaluate_model(y_test, y_pred3, \"Decision Tree with Min Split 20 and criterion as Gini\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:48.143151Z","iopub.execute_input":"2025-04-07T04:03:48.143647Z","iopub.status.idle":"2025-04-07T04:03:57.636540Z","shell.execute_reply.started":"2025-04-07T04:03:48.143610Z","shell.execute_reply":"2025-04-07T04:03:57.635098Z"}},"outputs":[{"name":"stdout","text":"Model: Standard Decision Tree\nAccuracy: 0.7423\nConfusion Matrix:\n[[14658  5177]\n [ 5044 14790]]\nClassification Report:\n              precision    recall  f1-score   support\n\n   Resistant       0.74      0.74      0.74     19835\n   Sensitive       0.74      0.75      0.74     19834\n\n    accuracy                           0.74     39669\n   macro avg       0.74      0.74      0.74     39669\nweighted avg       0.74      0.74      0.74     39669\n\n\n\nModel: Decision Tree with Max Depth 15 and criterion as Entropy\nAccuracy: 0.6282\nConfusion Matrix:\n[[16305  3530]\n [11217  8617]]\nClassification Report:\n              precision    recall  f1-score   support\n\n   Resistant       0.59      0.82      0.69     19835\n   Sensitive       0.71      0.43      0.54     19834\n\n    accuracy                           0.63     39669\n   macro avg       0.65      0.63      0.61     39669\nweighted avg       0.65      0.63      0.61     39669\n\n\n\nModel: Decision Tree with Min Split 20 and criterion as Gini\nAccuracy: 0.7448\nConfusion Matrix:\n[[14602  5233]\n [ 4890 14944]]\nClassification Report:\n              precision    recall  f1-score   support\n\n   Resistant       0.75      0.74      0.74     19835\n   Sensitive       0.74      0.75      0.75     19834\n\n    accuracy                           0.74     39669\n   macro avg       0.74      0.74      0.74     39669\nweighted avg       0.74      0.74      0.74     39669\n\n\n\n","output_type":"stream"}],"execution_count":16},{"cell_type":"code","source":"# Visualize the decision trees and save to files\ndef visualize_tree(dt_model, feature_names, class_names, max_depth=3, model_name=\"Decision Tree\"):\n    plt.figure(figsize=(40, 20))\n    plot_tree(dt_model, \n              max_depth=max_depth,\n              feature_names=feature_names,\n              class_names=class_names,\n              filled=True, \n              rounded=True,\n              fontsize=9)\n    plt.title(f\"{model_name} Visualization (Max Depth={max_depth})\")\n    plt.tight_layout()\n    \n    # Save to file\n    filename = f'/kaggle/working/tree_viz_{model_name.replace(\" \", \"_\")}.png'\n    plt.savefig(filename, dpi=300, bbox_inches='tight')\n    plt.close()\n    \n    # For more detailed visualization, also save using export_graphviz\n    try:\n        dot_data = export_graphviz(\n            dt_model,\n            max_depth=max_depth,\n            feature_names=feature_names,\n            class_names=class_names,\n            filled=True,\n            rounded=True,\n            special_characters=True,\n            out_file=None\n        )\n        # Save the dot file for later use\n        with open(f'/kaggle/working/tree_{model_name.replace(\" \", \"_\")}.dot', 'w') as f:\n            f.write(dot_data)\n    except Exception as e:\n        print(f\"Could not export dot file: {e}\")\n\n# Class names for visualization\nclass_names = ['Resistant', 'Sensitive']\n\n# Visualize the three trees\nvisualize_tree(dt1, processed_feature_names, class_names, model_name=\"Standard_DT\")\nvisualize_tree(dt2, processed_feature_names, class_names, model_name=\"DT_MaxDepth15_Entropy\")\nvisualize_tree(dt3, processed_feature_names, class_names, model_name=\"DT_MinSamplesSplit20_Gini\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:03:57.637659Z","iopub.execute_input":"2025-04-07T04:03:57.637965Z","iopub.status.idle":"2025-04-07T04:04:12.362125Z","shell.execute_reply.started":"2025-04-07T04:03:57.637938Z","shell.execute_reply":"2025-04-07T04:04:12.358461Z"}},"outputs":[],"execution_count":17},{"cell_type":"code","source":"# Feature importance analysis and save to files\ndef plot_feature_importance(dt_model, feature_names, model_name=\"Decision Tree\"):\n    # Get feature importances\n    importances = dt_model.feature_importances_\n    \n    # Create DataFrame for visualization - handle case where importances length doesn't match feature_names\n    indices = np.argsort(importances)[::-1]\n    top_indices = indices[:min(10, len(importances))]\n    \n    # Use only available features\n    valid_features = [feature_names[i] if i < len(feature_names) else f\"Feature_{i}\" for i in top_indices]\n    valid_importances = [importances[i] for i in top_indices]\n    \n    # Create DataFrame\n    importance_df = pd.DataFrame({\n        'Feature': valid_features,\n        'Importance': valid_importances\n    })\n    \n    # Plot\n    plt.figure(figsize=(12, 8))\n    sns.barplot(x='Importance', y='Feature', data=importance_df)\n    plt.title(f'Top 10 Feature Importances - {model_name}')\n    plt.tight_layout()\n    \n    # Save to file\n    filename = f'/kaggle/working/feature_importance_{model_name.replace(\" \", \"_\")}.png'\n    plt.savefig(filename, dpi=300, bbox_inches='tight')\n    plt.close()\n    \n    # Also save the feature importance data to CSV for reference\n    importance_df.to_csv(f'/kaggle/working/feature_importance_{model_name.replace(\" \", \"_\")}.csv', index=False)\n\n# Plot feature importance for all models\nplot_feature_importance(dt1, processed_feature_names, \"Standard_DT\")\nplot_feature_importance(dt2, processed_feature_names, \"DT_MaxDepth15_Entropy\")\nplot_feature_importance(dt3, processed_feature_names, \"DT_MinSamplesSplit20_Gini\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:04:12.364784Z","iopub.execute_input":"2025-04-07T04:04:12.365722Z","iopub.status.idle":"2025-04-07T04:04:14.393648Z","shell.execute_reply.started":"2025-04-07T04:04:12.365652Z","shell.execute_reply":"2025-04-07T04:04:14.392001Z"}},"outputs":[],"execution_count":18},{"cell_type":"code","source":"# Compare model performances and save to file\nmodel_names = ['Standard DT', 'DT with Max Depth 15 and criterion as Entropy', 'DT with Min Samples Split 20 and criterion as Gini']\naccuracies = [accuracy1, accuracy2, accuracy3]\n\nplt.figure(figsize=(10, 6))\nsns.barplot(x=model_names, y=accuracies)\nplt.xlabel('Model')\nplt.ylabel('Accuracy')\nplt.title('Comparison of Decision Tree Model Accuracies')\nplt.ylim(0, 1)\nplt.tight_layout()\n\n# Save accuracy comparison\nplt.savefig('/kaggle/working/model_accuracy_comparison.png', dpi=300, bbox_inches='tight')\nplt.close()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-07T04:04:14.394725Z","iopub.execute_input":"2025-04-07T04:04:14.395046Z","iopub.status.idle":"2025-04-07T04:04:14.809112Z","shell.execute_reply.started":"2025-04-07T04:04:14.395016Z","shell.execute_reply":"2025-04-07T04:04:14.807859Z"}},"outputs":[{"name":"stderr","text":"/usr/local/lib/python3.10/dist-packages/seaborn/_oldcore.py:1765: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.\n  order = pd.unique(vector)\n","output_type":"stream"}],"execution_count":19}]}